sql/core/src/main/scala/org/apache/spark/sql/execution/command/AnalyzeColumnCommand.scala

```diff
@@ -42,17 +42,28 @@ case class AnalyzeColumnCommand(
     if (tableMeta.tableType == CatalogTableType.VIEW) {
       throw new AnalysisException("ANALYZE TABLE is not supported on views.")
     }
-    val sizeInBytes = AnalyzeTableCommand.calculateTotalSize(sessionState, tableMeta)
+    val oldRowCount = tableMeta.stats.flatMap(_.rowCount.map(_.toLong)).getOrElse(-1L)
+    val oldColStats = tableMeta.stats.map(_.colStats).getOrElse(Map.empty)
+    val newSize = AnalyzeTableCommand.calculateTotalSize(sessionState, tableMeta)
 
     // Compute stats for each column
-    val (rowCount, newColStats) = computeColumnStats(sparkSession, tableIdentWithDB, columnNames)
+    val (newRowCount, colStats) = computeColumnStats(sparkSession, tableIdentWithDB, columnNames)
+
+    val newColStats = if (newRowCount == 0) {
+      // No need to keep column stats when there's no data.
+      Map.empty[String, ColumnStat]
+    } else if (newRowCount == oldRowCount) {
```

Comment:
Good point! So we need to change our strategy, should we always remove old column stats? I thought of this case because sometimes user wants add column stats for other columns.